File: swift/14_sorts/CountingSort.swift
//
//  CountingSort.swift
//  algo
//
//  Created by Wenru Dong on 2018/10/18.
//  Copyright © 2018年 Wenru Dong. All rights reserved.
//

import Foundation

// 假设数组中储存的都是非负整数
public func countingSort(_ a: inout [Int]) {
    if a.count <= 1 { return }
    
    var counts = Array(repeating: 0, count: a.max()! + 1)
    for num in a {
        counts[num] += 1
    }
    for i in 1..<counts.count {
        counts[i] = counts[i-1] + counts[i]
    }
    
    var aSorted = Array(repeating: 0, count: a.count)
    for num in a.reversed() {
        let index = counts[num] - 1
        aSorted[index] = num
        counts[num] -= 1
    }
    
    for (i, num) in aSorted.enumerated() {
        a[i] = num
    }
}

File: python/14_sorts/counting_sort.py
"""
    计数排序

    Author: Wenru
"""

from typing import List
import itertools

def counting_sort(a: List[int]):
    if len(a) <= 1: return
    
    # a中有counts[i]个数不大于i
    counts = [0] * (max(a) + 1)
    for num in a:
        counts[num] += 1
    counts = list(itertools.accumulate(counts))

    # 临时数组，储存排序之后的结果
    a_sorted = [0] * len(a)
    for num in reversed(a):
        index = counts[num] - 1
        a_sorted[index] = num
        counts[num] -= 1
    
    a[:] = a_sorted


if __name__ == "__main__":
    a1 = [1, 2, 3, 4]
    counting_sort(a1)
    print(a1)

    a2 = [1, 1, 1, 1]
    counting_sort(a2)
    print(a2)

    a3 = [4, 5, 0, 9, 3, 3, 1, 9, 8, 7]
    counting_sort(a3)
    print(a3)
